pkg/resolvers/k8s/watcher.go

```diff
@@ -54,119 +51,136 @@ func (w *watcher) Close() {
+	case change = <-changeCh:
 	}
 
-	// Translate kube api endpoint watch event to resolver address and put into map for easier lookup.
-	for _, subset := range event.Object.Subsets {
-		updatedAddresses, err := subsetToAddresses(w.target, subset)
+	// Translate kube api endpoint watch change to resolver address and put into map for easier lookup.
+	for _, subset := range change.Subsets {
+		var err error
+		endpointsToUpdate, err = subsetToAddresses(w.target, subset)
 		if err != nil {
-			return []*naming.Update(nil), errors.Wrap(err, "k8sresolver: failed to convert k8s endpoint subset to update Addr")
+			return []*naming.Update(nil), errors.Wrap(err, "failed to convert k8s endpoint subset to update Addr")
 		}
 
-		for _, address := range updatedAddresses {
-			updatedEndpoints[address] = struct{}{}
+		if len(endpointsToUpdate) > 0 {
+			// Expected port found.
+			break
 		}
+
+		// Target port not found yet. Maybe other subsets includes target one?
 	}
 
-	// Create updates to add new endpoints.
-	for addr, md := range updatedEndpoints {
-		if _, ok := w.lastUpdates[addr]; ok {
-			continue
+	switch change.typ {
+	case watch.Added:
+		// Create updates to add new endpoints.
+		for k, addr := range endpointsToUpdate {
+			_, ok := w.endpoints[k]
+			if ok {
+				return []*naming.Update(nil), errors.Errorf("malformed internal state for endpoints. "+
+					"On added event type, we got update for %v that already exists in %v. Doing resync...", k, w.endpoints)
+			}
+
+			w.endpoints[k] = addr
+			updates = append(updates, &naming.Update{Op: naming.Add, Addr: addr})
 		}
+	case watch.Modified:
+		for k, addr := range endpointsToUpdate {
+			oldAddr, ok := w.endpoints[k]
+			if !ok {
+				return []*naming.Update(nil), errors.Errorf("malformed internal state for endpoints. "+
+					"On modified event type, we got update for %v that does not exists in %v. Doing resync...", k, w.endpoints)
+			}
 
-		updates = append(updates, &naming.Update{Op: naming.Add, Addr: addr, Metadata: md})
-	}
-	// Create updates to delete old endpoints.
-	for addr := range w.lastUpdates {
-		if _, ok := updatedEndpoints[addr]; ok {
-			continue
+			updates = append(updates, &naming.Update{Op: naming.Delete, Addr: oldAddr})
+			w.endpoints[k] = addr
+			updates = append(updates, &naming.Update{Op: naming.Add, Addr: addr})
 		}
-		updates = append(updates, &naming.Update{Op: naming.Delete, Addr: addr, Metadata: nil})
-	}
+	case watch.Deleted:
+		// Create updates to delete old endpoints.
+		for k, addr := range endpointsToUpdate {
+			_, ok := w.endpoints[k]
+			if !ok {
+				return []*naming.Update(nil), errors.Errorf("malformed internal state for endpoints. "+
+					"On delete event type, we got update for %v that does not exists in %v. Doing resync...", k, w.endpoints)
+			}
 
-	w.lastUpdates = updatedEndpoints
+			updates = append(updates, &naming.Update{Op: naming.Delete, Addr: addr})
+			delete(w.endpoints, k)
+		}
+	}
```

Comment:
I think it worth to add `default` with some error log ("unknown event type").

Reply:
yea, I thought that in the beginning, but actually we check that in the streamer. I guess there is no harm in checking it here as well. Will do.